File: dropwizzard/cart-backend/src/main/java/com/mercadolibre/cart_backend/core/services/TaskLocatorService.java
package com.mercadolibre.cart_backend.core.services;

import com.mercadolibre.cart_backend.cart.tasks.addCartItemTask.AddCartItemTask_01_00_00;
import com.mercadolibre.cart_backend.cart.tasks.deleteCartItemTask.DeleteCartItemTask_01_00_00;
import com.mercadolibre.cart_backend.cart.tasks.deleteCartItemsTask.DeleteCartItemsTask_01_00_00;
import com.mercadolibre.cart_backend.cart.tasks.deleteCartTask.DeleteCartTask_01_00_00;
import com.mercadolibre.cart_backend.cart.tasks.getCartTask.GetCartTask_01_00_00;
import com.mercadolibre.cart_backend.cart.tasks.getCartTask.GetCartTask_02_00_00;
import com.mercadolibre.cart_backend.cart.tasks.updateCartItemTask.UpdateCartItemTask_01_00_00;
import com.mercadolibre.cart_backend.core.TaskBase;
import com.mercadolibre.cart_backend.core.entities.EntityBase;
import com.github.zafarkhaja.semver.Version;

import java.util.Map;

/**
 * Created by fvitali on 8/8/16.
 */
public class TaskLocatorService {
	public <E extends EntityBase> TaskBase<E> get(String taskName, String client, Version version, Map<String, Object> params) {

		switch (taskName) {
			case "getCart":
				if (version.lessThan(Version.valueOf("2.0.0"))) {
					return (TaskBase<E>) new GetCartTask_01_00_00();
				}
				return (TaskBase<E>) new GetCartTask_02_00_00();
			case "deleteCart":
				return (TaskBase<E>) new DeleteCartTask_01_00_00();
			case "addCartItem":
				return (TaskBase<E>) new AddCartItemTask_01_00_00();
			case "deleteCartItems":
				return (TaskBase<E>) new DeleteCartItemsTask_01_00_00();
			case "updateCartItem":
				return (TaskBase<E>) new UpdateCartItemTask_01_00_00();
			case "deleteCartItem":
				return (TaskBase<E>) new DeleteCartItemTask_01_00_00();
			default:
				return null;
		}
	}
}

File: node.typescript/cart-backend/src/core/services/taskLocatorService.ts
import Entity from '../contracts/entity';
import Task from '../contracts/task';
import GetCartTask from '../../cart/tasks/getCartTask/default/v01.00.00/getCartTask';
import DeleteCartTask_01_00_00 from '../../cart/tasks/deleteCartTask/deleteCartTask_01_00_00';
import AddCartItemTask_01_00_00 from '../../cart/tasks/addCartItemTask/addCartItemTask_01_00_00';
import DeleteCartItemsTask_01_00_00 from '../../cart/tasks/deleteCartItemsTask/deleteCartItemsTask_01_00_00';
import UpdateCartItemTask_01_00_00 from '../../cart/tasks/updateCartItemTask/updateCartItemTask_01_00_00';
import DeleteCartItemTask_01_00_00 from '../../cart/tasks/deleteCartItemTask/deleteCartItemTask_01_00_00';
import GetHomeWidgetsTask_01_00_00 from '../../home/tasks/getHomeWidgetsTask/getHomeWidgetsTask_01_00_00';
import GetHomeWidgetsTask_01_02_00 from '../../home/tasks/getHomeWidgetsTask/native/getHomeWidgetsTask_01_02_00';
import GetHomeWidgetsTask_01_02_01 from '../../home/tasks/getHomeWidgetsTask/native/android/getHomeWidgetsTask_01_02_01';
import DynamicClass from '../../cart/routes/dynamicClass'
import TaskVersion from '../../cart/routes/taskVersion'
import {TreeVersion, Node} from '../../cart/routes/treeVersion'

export default class TaskLocatorService {
    public get(taskName: string, apiClient: string, apiVersion: TaskVersion): any {

		// let taskMap: Map<string, Array<Map<string, string>>> = new Map();
		let treeVersion: TreeVersion = new TreeVersion();
        this.addVersionsToTree('src/cart/tasks/getCartTask', treeVersion);

		treeVersion.buildTree();
		treeVersion.findVersion(treeVersion.root, apiClient, apiVersion);

		// let tasksByClient: string = taskMap.get(apiClient);
		
		// [0].get(apiVersion.versionTwoDigits);


		return new DynamicClass('getCartTask_v01_00_00');

	}

	private addVersionsToTree(currentDirPath: string, treeVersion: TreeVersion): void {
        let fs = require('fs');
        let path = require('path');
        var self = this;
        fs.readdirSync(currentDirPath).forEach(function (name: any) {
            var filePath = path.join(currentDirPath, name);
            var stat = fs.statSync(filePath);
            if (stat.isFile()) {
				let dirname = path.dirname(filePath);
				let version = dirname.substr(dirname.lastIndexOf("/")+1, dirname.length).replace(/\./g,'_');
				let auxClient = dirname.substr(0, dirname.lastIndexOf("/"));
				let client = auxClient.substr(auxClient.lastIndexOf("/")+1, auxClient.length);
				let taskName = path.parse(filePath).name;
				if (taskName === 'getCartTask') {
					// let array = taskMap.get(client);

					let newVersion = new Map();
					newVersion.set(TaskVersion.getApiFormat(version), taskName+`_${version}`);
					treeVersion.addNode(new Node(newVersion, client));
					// if (!array) {
					// 	taskMap.set(client, [newVersion]);
					// } else {
					// 	array.push(newVersion);
					// }
				}
            } else if (stat.isDirectory()) {
                self.addVersionsToTree(filePath, treeVersion);
            }
        });
    }
}